src/main/kotlin/objects/BoardStateManager.kt

```diff
@@ -89,15 +89,128 @@ class BoardStateManager(
+        if (player1Died && player2Died) {
+            return false
+        }
+
+        if (player1Died) {
+            getConnection(Player.Player1).sendPacket(GameOverPacket(false))
+            getConnection(Player.Player2).sendPacket(GameOverPacket(true))
+            return true
+        }
+
+        if (player2Died) {
+            getConnection(Player.Player1).sendPacket(GameOverPacket(true))
+            getConnection(Player.Player2).sendPacket(GameOverPacket(false))
+            return true
+        }
+
+        return false
+    }
+
+    suspend fun withGameOverHandler(handler: suspend () -> Unit) {
+        handler()
+
+        if (this.boardState.deck_masters.all { (it?.run { health > 0 }) != false }) {
+            return
+        }
+
+        val player1 = Player.Player1
+        val player2 = Player.Player2
+
+        val player1Connection = getConnection(player1)
+        val player2Connection = getConnection(player2)
+
+        // TODO: These are temporary after we consider that deck_masters must exist
+        // TODO: Hence, when they do exist, remove the next few lines
+        val deckMasterPlayer1Opt = this.boardState.deck_masters[playerToIndex(Player.Player1)]
+        val deckMasterPlayer2Opt = this.boardState.deck_masters[playerToIndex(Player.Player2)]
+
+        if (temporarySpecialLogicWithGameOverHandler(deckMasterPlayer1Opt, deckMasterPlayer2Opt)) {
+            return
+        }
+
+        // NOTE: at this point, deckMasterPlayer1 & 2 are guaranteed to exist
+        val deckMasterPlayer1 = deckMasterPlayer1Opt!!
+        val deckMasterPlayer2 = deckMasterPlayer2Opt!!
+
+        val isPlayer1Empty =
+            isHandEmpty(player1) && isBoardEmptyForPlayerExceptFor(player1, deckMasterPlayer1) && isDeckEmptyForPlayer(
+                player1
+            )
+        val isPlayer2Empty =
+            isHandEmpty(player2) && isBoardEmptyForPlayerExceptFor(player2, deckMasterPlayer2) && isDeckEmptyForPlayer(
+                player2
+            )
+
+        if (isPlayer1Empty && isPlayer2Empty) {
+            if (deckMasterPlayer1.health == deckMasterPlayer2.health) {
+                // tie
+                val winner = arrayOf(player1, player2).random()
+                print("tie condition")
+                print(winner.name)
+                getConnection(winner).sendPacket(
+                    GameOverPacket(true),
+                )
+
+                getConnection(!winner).sendPacket(
+                    GameOverPacket(false),
+                )
+                return
+            } else {
+                player1Connection.sendPacket(
+                    GameOverPacket(deckMasterPlayer1.health > deckMasterPlayer2.health),
+                )
+
+                player2Connection.sendPacket(
+                    GameOverPacket(deckMasterPlayer1.health < deckMasterPlayer2.health),
+                )
+            }
+
+            // TODO: is there a better way to close the connection?
```

Comment:
There is a `DisconnectPacket`, but that is more for unexpected disconnects like auth being invalid or the protocol being too old. If we don't use that, then this is OK as long as nothing uses the connection after.

Reply:
Okay. I think I'll keep this method because it makes more sense for the server to forcefully drop the connection here.